src/aleph/web/controllers/storage.py

```diff
@@ -56,20 +92,129 @@ async def add_storage_json_controller(request: web.Request):
+        item_content = json.loads(message.item_content)
+    actual_item_hash = sha256(content).hexdigest()
+    client_item_hash = item_content["item_hash"]
+    if len(content) > (1000 * MiB):
+        raise web.HTTPRequestEntityTooLarge(
+            actual_size=len(content), max_size=(1000 * MiB)
+        )
+    elif actual_item_hash != client_item_hash:
+        raise web.HTTPUnprocessableEntity()
+
+
+class StorageMetadata(pydantic.BaseModel):
+    message: PendingStoreMessage
+    file_size: int
+    sync: bool
+
+
+async def storage_add_file_with_message(
+    request: web.Request, session: DbSession, chain_service, storage_metadata, file_io
+):
+    config = get_config_from_request(request)
+    mq_queue = None
+
+    pending_store_message = PendingStoreMessage.parse_obj(storage_metadata.message)
+    pending_message_db = PendingMessageDb.from_obj(
+        obj=pending_store_message, reception_time=dt.datetime.now(), fetched=True
+    )
+    await _verify_message_signature(
+        pending_message=storage_metadata.message, chain_service=chain_service
+    )
+    await _verify_user_balance(
+        session=session,
+        pending_message_db=pending_message_db,
+        size=storage_metadata.file_size,
+    )
+    await _verify_user_file(
+        message=storage_metadata.message,
+        size=storage_metadata.file_size,
+        file_io=file_io,
+    )
+
+    if storage_metadata.sync:
+        mq_channel = await get_mq_channel_from_request(request, logger=logger)
+        mq_queue = await mq_make_aleph_message_topic_queue(
+            channel=mq_channel,
+            config=config,
+            routing_key=f"*.{pending_message_db.item_hash}",
+        )
+
+    session.add(pending_message_db)
+    session.commit()
+    if mq_queue:
+        mq_message = await mq_read_one_message(mq_queue, 30)
+
+        if mq_message is None or mq_message.routing_key is None:
+            raise web.HTTPAccepted()
+
+        status_str, _item_hash = mq_message.routing_key.split(".")
+        processing_status = MessageProcessingStatus(status_str)
+        status_code = processing_status_to_http_status(processing_status)
+        return web.json_response(status=status_code, text=_item_hash)
+
+
 async def storage_add_file(request: web.Request):
     storage_service = get_storage_service_from_request(request)
     session_factory = get_session_factory_from_request(request)
-
-    # No need to pin it here anymore.
-    # TODO: find a way to specify linked ipfs hashes in posts/aggr.
+    chain_service: ChainService = get_chain_service_from_request(request)
     post = await request.post()
     file_io = multidict_proxy_to_io(post)
+    post = await request.post()
+    metadata = post.get("metadata", b"")
+    storage_metadata = None
+    try:
+        if isinstance(metadata, FileField):
+            metadata_content = metadata.file.read()
+            storage_metadata = StorageMetadata.parse_raw(metadata_content)
+        else:
```

Comment:
I'm pretty sure this `else` breaks the endpoint when the metadata is not specified. You default to `b""` so you will get a Pydantic exception in this branch.